### PR TITLE
Add MXFP4 (including Quark W4A4) quantization support for DeepSeek-architecture on ROCm

### PR DESCRIPTION
**Pull request by JohnQinAMD**
Enable MXFP4  (including Quark W4A4) quantized models (e.g. GLM-5-MXFP4-Q8, Kimi-K2.5-MXFP4) to load and run on AMD MI355X with ROCm. These models use Quark's W4A4 MXFP4 scheme with block-scaled FP4 for MoE expert weights and BF16 for attention.

Key changes:

mxfp4.py:
- Return None instead of raising NotImplementedError for attention layers (attention uses BF16 weights, not MXFP4)

fused_moe_triton/layer.py:
- Skip weight transpose for packed uint8 (MXFP4) tensors during loading
- Pre-compute expert_mask_gpu in __init__ for CUDA graph compatibility
- Fix expert_id None check (was `not expert_id` which fails for id=0)

quark/:
- Fix Quark MXFP4 MoE scheme weight creation and loading for block-scaled format with proper shard handling
- Add per-expert weight processing support

deepseek_v2.py:
- Set packed_modules_mapping for Quark weight loader compatibility
- Disable shared expert fusion for MXFP4 serialized models (routed experts are quantized but shared experts remain BF16)
- Disable shared expert fusion for Quark models (mixed precision)

deepseek_weight_loader.py:
- Handle fused_qkv_a_proj_with_mqa weight mapping for Quark checkpoints

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4_moe.py [hunk 1]
```diff
@@ -206,12 +247,12 @@ def apply_weights(
             quant_type=QuantType.per_1x32,
             w1_scale=layer.w13_weight_scale,
             w2_scale=layer.w2_weight_scale,
-            activation=(
-                ActivationType.Silu
-                if moe_runner_config.activation == "silu"
-                else ActivationType.Gelu
-            ),
-            doweight_stage1=False,
+            activation=ActivationType.Swiglu,
```
Comment (BowenBao) [inline python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4_moe.py:250]: is this correct to fix as swiglu?

File: python/sglang/srt/layers/quantization/quark/utils.py [hunk 2]
```diff
@@ -125,7 +125,9 @@ def _is_equal_or_regex_match(
 def b_dynamic_mxfp4_quant(x):
     h, b, d = x.shape
     x, x_scales = dynamic_mxfp4_quant(x.reshape(-1, d))
-    return x.view(h, b, d // 2), x_scales.view(h, b, d // 32)
+    x_out = x.view(h, b, d // 2).contiguous()
```
Comment (BowenBao) [inline python/sglang/srt/layers/quantization/quark/utils.py:128]: are these necessary changes?

File: python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4_moe.py [hunk 3]
```diff
@@ -133,35 +130,85 @@ def create_weights(
         layer.register_parameter("w13_weight_scale", w13_weight_scale)
         layer.register_parameter("w2_weight_scale", w2_weight_scale)
 
+        # WEIGHT_BIAS (zeros, matching standard mxfp4 path for CUDA graph compat)
+        w13_weight_bias = torch.nn.Parameter(
```
Comment (BowenBao) [inline python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4_moe.py:134]: not all models have biases, could we make this conditional?

File: python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4_moe.py [hunk 4]
```diff
@@ -133,35 +130,85 @@ def create_weights(
...
+            requires_grad=False,
+        )
+        layer.register_parameter("w2_weight_bias", w2_weight_bias)
+        set_weight_attrs(w2_weight_bias, extra_weight_attrs)
+
+        self.hidden_pad = 0
+        self.intermediate_pad = 0
+
     def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
-        float_dtype = torch.get_default_dtype()
```
Comment (BowenBao) [inline python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4_moe.py:137]: I'm not sure about the shuffling as well as activation changes.

**Review by BowenBao — CHANGES_REQUESTED** [review]
The deepseek mxfp4 models, such as https://huggingface.co/amd/DeepSeek-R1-MXFP4 and https://huggingface.co/amd/DeepSeek-R1-0528-MXFP4, are well supported by SGLang. I'm not sure what additional functionality this PR is intended to provide.
